server/pages/index.js

```diff
@@ -342,13 +342,74 @@ function setUpLoggedInRoute( req, res, next ) {
 	}
 }
 
-function setUpRoute( req, res, next ) {
-	if ( req.cookies.wordpress_logged_in ) {
-		// the user is probably logged in
-		setUpLoggedInRoute( req, res, next );
-	} else {
-		setUpLoggedOutRoute( req, res, next );
+/**
+ * Sets up a Content Security Policy header
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+ * @param {Object} req Express request object
+ * @param {Object} res Express response object
+ * @param {Function} next a callback to call when done
+ */
+function setUpCSP( req, res, next ) {
+	if ( ! req.originalUrl.match( /^\/log-in\// ) ) {
+		next();
+		return;
 	}
+
+	const inlineScripts = [
+		'sha256-3yiQswl88knA3EhjrG5tj5gmV6EUdLYFvn2dygc0xUQ=',
+		'sha256-ZKTuGaoyrLu2lwYpcyzib+xE4/2mCN8PKv31uXS3Eg4=',
+	];
+
+	req.context.inlineScriptNonce = crypto.randomBytes( 48 ).toString( 'hex' );
+	req.context.analyticsScriptNonce = crypto.randomBytes( 48 ).toString( 'hex' );
+
+	const policy = {
+		'default-src': [ "'self'" ],
+		'script-src': [
+			"'self'",
+			"'report-sample'",
+			"'unsafe-eval'",
+			'stats.wp.com',
+			'https://apis.google.com',
+			`'nonce-${ req.context.inlineScriptNonce }'`,
+			`'nonce-${ req.context.analyticsScriptNonce }'`,
+			...inlineScripts.map( hash => `'${ hash }'` ),
+		],
+		'base-uri': [ "'none'" ],
+		'style-src': [ "'self'", '*.wp.com', 'https://fonts.googleapis.com' ],
+		'form-action': [ "'self'" ],
+		'object-src': [ "'none'" ],
+		'img-src': [ "'self'", '*.wp.com', 'https://www.google-analytics.com' ],
+		'frame-src': [ "'self'", 'https://public-api.wordpress.com', 'https://accounts.google.com/' ],
+		'font-src': [
+			"'self'",
+			'*.wp.com',
+			'https://fonts.gstatic.com',
+			'data:', // should remove 'data:' ASAP
+		],
+		'media-src': [ "'self'" ],
+		'connect-src': [ "'self'", 'https://wordpress.com/' ],
+		'report-uri': [ '/cspreport' ],
+	};
+
+	const policyString = Object.keys( policy )
+		.map( key => `${ key } ${ policy[ key ].join( ' ' ) }` )
+		.join( '; ' );
+
+	res.set( { 'Content-Security-Policy-Report-Only': policyString } );
```

Comment:
I see, we're in report only mode... We should still add some comments in the code to make that more obvious when reading the code